Greatly reduce AI lag

DizzyEggg opened the pull request:
This was not visible in single battles, but there was a slight delay in double battles and a long 1 sec delay in double partner battles, caused by the AI thinking logic.

The AI calculations are expensive, because the same code has to run for each battler's move against each target. Which means we have to be careful with what operations are done, and if possible only run the heavy ones once. It goes without saying that agbcc performs much worse than modern.
With #3302 merged, we could now easily measure the lag.

Some benchmarks(flags: No Bad Moves, Viability, Try to Faint) to illustrate the issue: (agbbc - modern)
- singles - 11 - 10 - here it was not so bad, barely noticeable
![agbcc_single_old](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/33cd1dc3-726d-4164-9237-0b57f5f942e1)
![modern_single_old](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/42ed3fba-c47e-4c70-932b-d93c6c337ed3)
- doubles - 40 - 26 - here it was really bad for agbcc, and slightly bad for modern
![agbcc_double_old](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/963d72f7-d412-4b06-bd07-eaf8465b5f6a)
![modern_double_old](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/486abbb4-ff71-4cfe-93fa-9838250eaac5)
- doubles with AI partner - 62 - 40 - 1sec+ delay for agbcc which is horrendous, for modern it was _only_ as bad as doubles for agbcc, which shows how bad agbcc is
![agbcc_triple_old](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/4686458b-cb9e-43f3-b0fd-7bcc039857a0)
![modern_triple_old](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/3baec957-5b1c-49ac-896c-cb71a7802801)

I did a couple of various optimization in the AI and battle code to reduce the lag, and I'm quite happy with the results. 
New benchmarks: (agbbc - modern)
- singles - 7 - 7 - It's sometimes 6 for modern and 8 for agbcc, but mostly 7 for both. Looks like it's squeezed to the max here
![agbcc_single_new](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/a8a334db-8ce9-40d3-81db-8d06526c5f4c)
![modern_single_new](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/d7814954-0412-4792-a09e-021b42e848c9)
- doubles - 19 - 12 - The real killer is here, modern on doubles is now almost as fast as singles on agbcc before the fix. 19 is okayish, but it can be noticed, so to the trash can agbcc goes
![agbcc_double_new](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/ba971cd3-1b33-4bdf-9b1f-bb0c879e493e)
![modern_double_new](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/7567f398-3fe2-405e-9a82-da5ee46b2aa9)
- doubles with AI partner - 33 - 22 - Modern here is faster than doubles on agbcc before the fix. 
![agbcc_triple_new](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/0e1fa4a0-b969-4b5a-9bac-afb760c858b3)
![modern_triple_new](https://github.com/rh-hideout/pokeemerald-expansion/assets/16259973/0d685c97-2364-4271-a25c-7f22e228066c)

Obviously, these may sligthly change, as the lag depends on the battlers' moves, abilities, mons count, etc. But it should be somewhere around that value.

What's changed:
- DoMoveDamageCalc was heavily optimized by inlining pretty much all of the calls, and by adding new local variables, so that there are not any additional calls to functions like `GetBattlerAbility` or `GetBattlerHoldEffect`. 
- I modified `AI_WhoStrikesFirst` and `GetWhoStrikesFirst` for the same reason, to reduce the function calls and reuse the abilities/hold effects it sets at the beginning. Also, adds a new Ai Data `speedStats` variable, so it calculates speed only once
- New Ai Data var `weatherHasEffect` to reduce the calls to `WEATHER_HAS_EFFECT`. New function which uses the var AI_GetWeather
- Lots of local u8/u16/s8/bool8 vars, etc. were changed to u32 and 32, so the generated code doesn't bother with lsl/lsr to make sure the variables match the size
- Simplified BattleAI_DoAIProcessing
- AI_DATA and AI_THINKING_STRUCT are now sometimes local variables, so the functions don't have to load `gBattleResources` pointers every time they're accessed.
- SetAiLogicDataForTurn modified. Also there was a bug before in that the dmgs were not simulated for player mons, which most likely resulted in buggy code if AI wanted to know how much dmg player can deal. So now it runs even more damage calcs, and yet is much faster. I'd say it's a success.
- Crit dmg is not calculated every time. Again, damage calc is _costly_. Now ai only calculates dmg for a crit show if it's at least +1 crit stage.
- Some other small modifications and optimizations. The general idea is the same, reduce the number of calls, use the variables set in the Ai Logic Data struct.

Important note:
This doesn't add any new AI logic, it just makes the same code(well, in theory) run faster. (Though there could be some 'accidental' bug fixes).

ghoulslash commented on src/battle_util.c:
```diff
@@ -6313,7 +6313,7 @@ bool32 IsBattlerTerrainAffected(u8 battler, u32 terrainFlag)
 }
 
-bool32 CanSleep(u8 battler)
+bool32 CanSleep(u32 battler)
```
would updating these CanBeStatused functions to take ability, hold effect etc speed up AI calcs?

DizzyEggg replied:
I think so, tho not sure if it would be enough to get a single frame.

ghoulslash reviewed (commented):
still working on getting some testing done but wanted to send some initial thoughts

BuffelSaft commented:
Personally I would scrap the crit damage calc entirely unless it's a guaranteed crit, or at the very least +2. As it stands it's a lot of wasted time on something with a 12.5% chance to happen, and for little to no benefit to the AI. Situations where that calc makes it dumber are more common than situations where it helps, and situations where it might help can be handled with basic checks (are screens up, is target's defense stat boosted) instead. 

E.g. if the AI is going for a crucial KO that it will just barely miss, but thinks it will get due to the extra damage from factoring in crit damage (and the things it ignores), it may go for the high crit move when a slightly stronger move would have been a guaranteed KO, and the crit move needs a crit. This makes all of the "can battler KO battler" unreliable.

We don't account for miss chance (unless that was changed?) - which can be as high as 70% - so why account for 12.5% crit chances?

DizzyEggg commented:
> Personally I would scrap the crit damage calc entirely unless it's a guaranteed crit, or at the very least +2. As it stands it's a lot of wasted time on something with a 12.5% chance to happen, and for little to no benefit to the AI. Situations where that calc makes it dumber are more common than situations where it helps, and situations where it might help can be handled with basic checks (are screens up, is target's defense stat boosted) instead.
> 
> E.g. if the AI is going for a crucial KO that it will just barely miss, but thinks it will get due to the extra damage from factoring in crit damage (and the things it ignores), it may go for the high crit move when a slightly stronger move would have been a guaranteed KO, and the crit move needs a crit. This makes all of the "can battler KO battler" functions unreliable.
> 
> We don't account for miss chance (unless that was changed?) - which can be as high as 70% - so why account for 12.5% crit chances?

I was thinking about the +2 threshold too. Though please note that if the crit stage is only +1, the AI won't treat the damage as if it was the crit one, instead it will be something more like 1,1x normal dmg.

Accounting for accuracy is done in this PR #3199

BuffelSaft commented:
> I was thinking about the +2 threshold too. Though please note that if the crit stage is only +1, the AI won't treat the damage as if it was the crit one, instead it will be something more like 1,1x normal dmg.
> 
> Accounting for accuracy is done in this PR #3199

Yeah, but that extra 10% damage can really mess up the decision making when a lot of it is based on HP thresholds rather than average damage. If you've got multiple moves that can KO, but one move says it does 10% more damage than it actually does, and it won't get the KO without that 10%, it shouldn't be treated as if it will KO, because 87.5% of the time it won't.

I've had a few comments like "why did the AI go for Night Slash when Drill Peck was a guaranteed KO?" and this was the culprit. In that situation, Night Slash +10% damage was a definite KO, and so was Drill Peck, but really Night Slash without that extra 10% was not a KO (or at least not a high chance). That led to the AI thinking both moves were equally good choices, so it chose one randomly and got Night Slash. IMO it should have seen that Drill Peck was the better option here and gone for that.

BuffelSaft commented:
> In this case I wouldn't mess with +2 because the 20% boost (assuming it's 1.2x) will mess it up even more. Make it only if it's guaranteed.

Yeah, though at least getting the crit is a coin toss in that case, so it might at least mess with players trying to predict the AI if it randomly goes for the weaker move half the time. Probably still better just to go for the KO though, especially if the AI is faster

ghoulslash commented:
Have you done any comparisons with not running WEATHER_HAS_EFFECT and storing move damage results up front? From my own testing I kept seeing reduced performance as I added more things to the calc damage routines, because weather and move damage results are not always used, so we're increasingly the baseline computation time to handle specific future calculations that aren't always needed. Granted, my personal implementation is more complex due to ability arrays and field damping stuff, but I'm curious if you did a comparison when including weather up front or not.

Also, a minor bug I found in the initial simulatedDmg loop is that we should be using GetMovesArray, because the AI should only simulate dmg on known player moves.

DizzyEggg commented:
> Have you done any comparisons with not running WEATHER_HAS_EFFECT and storing move damage results up front? From my own testing I kept seeing reduced performance as I added more things to the calc damage routines, because weather and move damage results are not always used, so we're increasingly the baseline computation time to handle specific future calculations that aren't always needed. Granted, my personal implementation is more complex due to ability arrays and field damping stuff, but I'm curious if you did a comparison when including weather up front or not.
> 
> Also, a minor bug I found in the initial simulatedDmg loop is that we should be using GetMovesArray, because the AI should only simulate dmg on known player moves.

Yes, it's faster, because weather is **always** checked in the damage calc in `GetWeatherDamageModifier`. So that's two function calls reduced to just a simple if true check. 
For the damage calc make sure the functions are inline. 

You're also right about the bug, I fixed it in this PR.

ghoulslash commented:
GetWeatherDamageModifier should onlyn be running  WEATHER_HAS_EFFECT if gBattleWeather is non-zero. So it's sort of a trade between always running WEATHER_HAS_EFFECT up front vs. waiting until there there actual weather and then it gets run for all damage calc calls.
But it looks like GetWeather checks gBattleWeather being 0 first so I think that should be fine.

The only other comment I had is that I blindly converted a ton of variables to u32 and saw some weird battle effects so ended up scrapping that as well, so we should be testing this PR pretty rigorously unfortunately. The weird thing I saw was an assertion that battler<gBattlersCount in PrepareDataTransfer
